### PR TITLE
Workaround Ubuntu kernel oops

### PR DESCRIPTION
**Pull request by schu**
Workaround for https://github.com/weaveworks/scope/issues/3131 until the issue is fixed upstream.

Ubuntu bug report: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454

### CODE REVIEW

File: probe/endpoint/ebpf.go [hunk 1]
```diff
@@ -84,6 +84,31 @@ func isKernelSupported() error {
...
+		// cut away upload number
+		abiNumberStr := strings.Split(vendorStr, ".")[0]
+		abiNumber, err := strconv.Atoi(abiNumberStr)
+		if err != nil {
+			return err
+		}
+		// TODO: give the check an upper limit once the bug is fixed
+		if abiNumber >= 119 {
+			// https://github.com/weaveworks/scope/issues/3131
+			// https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454
```
Comment (alban) [inline probe/endpoint/ebpf.go]: I'm guessing the fix in the Ubuntu kernel is to be released soon. Should we wait until it is released, so we know the exact abiNumber range where it's broken?

File: probe/endpoint/ebpf.go [hunk 2]
```diff
@@ -84,6 +84,31 @@ func isKernelSupported() error {
...
+
+		// The vendor specific release part could look like:
+		// '<ABI number>.<upload number>-<flavour>' or
+		// '<ABI number>-<flavour>'
+		// See https://wiki.ubuntu.com/Kernel/FAQ
+
+		// cut away flavour
+		vendorStr := strings.Split(releaseParts[4], "-")[0]
+		// cut away upload number
+		abiNumberStr := strings.Split(vendorStr, ".")[0]
```
Comment (bboreham) [inline probe/endpoint/ebpf.go]: I had to read the docs for `Split` to see what it did if the separator isn't found.
Why not just include the pattern for the ABI number in `releaseRegex`, i.e. replace `(.*)$` with `(\d*)(.*)$` ?

File: probe/endpoint/ebpf_test.go [hunk 3]
```diff
@@ -233,3 +235,59 @@ func TestInvalidTimeStampDead(t *testing.T) {
...
+		{
+			"4.4.0-119-generic",
+			"#143-Ubuntu SMP Mon Apr 2 16:08:24 UTC 2018",
+			false,
+		},
+		{
+			"4.4.0-116-generic",
+			"#140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018",
+			true,
+		},
```
Comment (bboreham) [inline probe/endpoint/ebpf_test.go]: The tests are all for older Ubuntus; can I suggest adding some others? For example:
* The newest Ubuntu in the list on GCP - `4.13.0-38-generic #43-Ubuntu SMP Wed Mar 14 15:20:44 UTC 2018`
* A newer Ubuntu with a 119 abi - `4.13.0-119-generic #43-Ubuntu SMP Wed Apr 1 00:00:00 UTC 2018`
* Default Debian on GCP - `4.9.0-6-amd64 #1 SMP Debian 4.9.82-1+deb9u3 (2018-03-02)`

File: probe/endpoint/ebpf.go (hunk 1 above)
Comment (rade) [reply probe/endpoint/ebpf.go]: > I'm guessing the fix in the Ubuntu kernel is to be released soon.

Any idea how long that might take?

Frankly, if there is going to be a new LTS security-updates kernel release with the fix in the next few days, I am not sure it's even worth bothering having any check here at all.

File: probe/endpoint/ebpf_test.go (hunk 3 above)
Comment (schu) [reply probe/endpoint/ebpf_test.go]: Added, plus tightened the check in `isKernelSupported` to only match minor version `4`. Thanks @bboreham

File: probe/endpoint/ebpf.go (hunk 1 above)
Comment (schu) [reply probe/endpoint/ebpf.go]: >> I'm guessing the fix in the Ubuntu kernel is to be released soon.

> Any idea how long that might take?

We can't tell and have asked Seth (Ubuntu) if he can give an ETA.

> Frankly, if there is going to be a new LTS security-updates kernel release with the fix in the next few days, I am not sure it's even worth bothering having any check here at all.

I agree.

Comment (schu) [reply probe/endpoint/ebpf.go]: Update: approximately around 4 weeks for the update to be published, https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454/comments/6

File: probe/endpoint/ebpf.go [hunk 4]
```diff
@@ -84,6 +96,22 @@ func isKernelSupported() error {
 		return fmt.Errorf("got kernel %s but need kernel >=4.4", release)
 	}
 
+	if strings.Contains(version, "Ubuntu") {
+		// Check for specific Ubuntu kernel versions with
+		// known issues.
+
+		abiNumber, err := strconv.Atoi(releaseParts[4])
+		if err != nil {
+			return err
```
Comment (bboreham) [inline probe/endpoint/ebpf.go]: I think this should return nil, on the basis it's at least 4.4 and we don't know that it is 119-ish, so just allow it.

Comment (schu) [reply probe/endpoint/ebpf.go]: Updated, thanks.

**Pull request merged**